packages/react-pages/src/node/dynamic-modules/PageStrategy.ts

```diff
@@ -0,0 +1,282 @@
+import * as fs from 'fs-extra'
+import * as path from 'path'
+import { EventEmitter } from 'events'
+import chokidar, { FSWatcher } from 'chokidar'
+import { debounce } from 'mini-debounce'
+import { PagesData } from './pages'
+import {
+  extractStaticData,
+  defaultPageLoader,
+  defaultPageFinder,
+} from './utils'
+
+export class PageStrategy extends EventEmitter {
+  readonly getPages: () => Promise<Readonly<PagesData>>
+  readonly close: () => void
+
+  constructor(
+    pagesDir: string,
+    findPages: FindPages = defaultPageFinder,
+    loadPageData: LoadPageData = defaultPageLoader
+  ) {
+    super()
+
+    const pageCache: PagesData = {}
+    const fileCache: FileCache = {}
+    const watchers = new Set<FSWatcher>()
+
+    // The "promise" event is for HMR updates to the pages map.
+    const emitPromise = debounce(() => {
+      this.emit('promise', pagesPromise)
+    }, 100)
+
+    // The "change" event is for HMR updates to each page.
+    const changedPages = new Set<string>()
+    const emitChange = debounce(() => {
+      this.emit('change', Array.from(changedPages))
+      changedPages.clear()
+    }, 100)
+
+    const setPageData = (pageId: string, pageData: Partial<PageData>) => {
```

Comment:
I prefer move big functions (like this) after the return statement (as function declarations will host). So that the directly-executed code will be concise. (We don't need to read this function in the first reading)